pmlc/dialect/pxa/transforms/stencil.cc

```diff
@@ -107,33 +121,76 @@ mlir::Optional<GemmOperationMatch> Stencil::getGemmOperation() {
     // Now verify the types of the operands of the mulOp must be affine.load operations.
     if (!lhs || !rhs) {
       IVLOG(3, "the lhs or rhs of the mul operation are not an affne.load operations");
+      ret = false;
       return;
     }
 
     // TODO: Need a bit better liveness analysis here to make sure the parameters 
     // of any of the above 4 operations are not used in operations with
     // side effects - store, calls, etc.
 
+    if (mulOpType == NoneMulOpType ||
+        !in1Op ||
+        !in2Op ||
+        !outOp) {
+      ret = false;
+      return;
+    }
+
     // Fill the values for the in/out/type of multiplication, etc.
-    GemmOperationMatch gemmMatch = GemmOperationMatch();
-    gemmMatch.mulOpType = mulOpType;
-    gemmMatch.in1Op = lhs;
-    gemmMatch.in2Op = rhs;
-    gemmMatch.outOp = reduceOp;
-    ret = gemmMatch;
-    op.setAttr("is_gemm", builder.getUnitAttr());
+    this->mulOpType = mulOpType;
+    in1Op = lhs;
+    in2Op = rhs;
+    outOp = reduceOp;
+    ret = true;
   });
 
   return ret;
 }
 
+bool Stencil::CollectTensors() {
+  assert(in1Op && in2Op && outOp);
+
+  tensors.push_back(in1Op.getMemRef());
+  tensors.push_back(in2Op.getMemRef());
+  tensors.push_back(outOp.out());
+
+  return tensors.size() == kNumTensors;
+}
+
+bool Stencil::ComputeStrideInfo() {
+  assert(in1Op && in2Op && outOp);
+
+  auto in1OpOptional = computeStrideInfo(in1Op);
+  auto in2OpOptional = computeStrideInfo(in2Op);
+  auto outOpOptional = computeStrideInfo(outOp);
+
+  if (!in1OpOptional ||
+      !in2OpOptional ||
+      !outOpOptional)
+    return false;
+
+  tensorsStrides.push_back(*in1OpOptional);
+  tensorsStrides.push_back(*in2OpOptional);
+  tensorsStrides.push_back(*outOpOptional);
+  return tensorsStrides.size() == kNumTensors;
+}
+
 void Stencil::DoStenciling() {
   // Initialization
-  auto matchOpt = getGemmOperation();
-  if (!matchOpt.hasValue()) {
+  if (!TryIdentifyGemmOperation()) {
     IVLOG(3, "Not a Gemm match.");
     return;
   }
+
+  if (!CollectTensors())
+    return;
+
+  if (!ComputeStrideInfo())
+    return;
+
+  Builder builder(op.getContext());
+  op.setAttr("is_gemm", builder.getUnitAttr());
```

Comment:
There's no reason to create a builder just to create a `UnitAttr`. You can call `UnitAttr::get()`

Reply:
Not getting the builder anymore and using the static factory.